data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/InjectionRangeActionAdderImpl.java

```diff
@@ -0,0 +1,102 @@
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+import static com.farao_community.farao.data.crac_impl.AdderUtils.assertAttributeNotEmpty;
+import static com.farao_community.farao.data.crac_impl.AdderUtils.assertAttributeNotNull;
+
+/**
+ * @author Baptiste Seguinot {@literal <baptiste.seguinot at rte-france.com>}
+ */
+public class InjectionRangeActionAdderImpl extends AbstractStandardRangeActionAdder<InjectionRangeActionAdder> implements InjectionRangeActionAdder {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(InjectionRangeActionAdderImpl.class);
+
+    private final List<DistributionKeyOnNetworkElement> distributionKeys;
+
+    @Override
+    protected String getTypeDescription() {
+        return "InjectionRangeAction";
+    }
+
+    InjectionRangeActionAdderImpl(CracImpl owner) {
+        super(owner);
+        distributionKeys = new ArrayList<>();
+    }
+
+    @Override
+    public InjectionRangeActionAdder withNetworkElementAndKey(double key, String networkElementId) {
+        return withNetworkElementAndKey(key, networkElementId, networkElementId);
+    }
+
+    @Override
+    public InjectionRangeActionAdder withNetworkElementAndKey(double key, String networkElementId, String networkElementName) {
+        distributionKeys.add(new DistributionKeyOnNetworkElement(networkElementId, networkElementName, key));
+        return this;
+    }
+
+    @Override
+    public InjectionRangeAction add() {
+        checkId();
+        if (!Objects.isNull(getCrac().getRemedialAction(id))) {
+            throw new FaraoException(String.format("A remedial action with id %s already exists", id));
+        }
+
+        // check network elements
+        checkNetworkElements();
+        assertAttributeNotEmpty(distributionKeys, "InjectionRangeAction", "injection distribution key", "withNetworkElementAndKey()");
+
+        // check ranges
+        assertAttributeNotEmpty(ranges, "InjectionRangeAction", "range", "newRange()");
+
+        // check usage rules
+        if (usageRules.isEmpty()) {
+            LOGGER.warn("InjectionRangeAction {} does not contain any usage rule, by default it will never be available", id);
+        }
+
+        Map<NetworkElement, Double> neAndDk = addNetworkElements();
+        InjectionRangeAction injectionRangeAction = new InjectionRangeActionImpl(this.id, this.name, this.operator, this.groupId, this.usageRules, this.ranges, neAndDk);
+        this.getCrac().addInjectionRangeAction(injectionRangeAction);
+        return injectionRangeAction;
+    }
+
+    private void checkNetworkElements() {
+        distributionKeys.forEach(dK -> assertAttributeNotNull(dK.networkElementId, "InjectionRangeAction", "network element", "withNetworkElementAndKey()"));
+    }
+
+    private Map<NetworkElement, Double> addNetworkElements() {
+        Map<NetworkElement, Double> distributionKeyMap = new HashMap<>();
+        distributionKeys.forEach(sK -> {
+            if (sK.distributionKey != 0.0) {
```

Comment:
check if abs < epsilon

Reply:
also, what would happen if the value was NaN?

Reply:
if check if abs < epsilon

cannot be NaN according to intelliJ. Maybe because it a double, and not a Double (?).

anyway, now i put `Math.abs(sK.distributionKey) > 1e-3`, which will return false if distributionKey is NaN.